src/server/game/AI/ScriptedAI/ScriptedEscortAI.cpp

```diff
@@ -384,9 +366,41 @@ void npc_escortAI::OnPossess(bool apply)
 
 void npc_escortAI::AddWaypoint(uint32 id, float x, float y, float z, uint32 waitTime)
 {
-    Escort_Waypoint t(id, x, y, z, waitTime);
+    int32 entry = -int32(me->GetEntry());
 
-    WaypointList.push_back(t);
+    WaypointPathContainer::iterator itr = sWaypointMgr->GetWaypointPathContainer().find(entry);
+    if (itr != sWaypointMgr->GetWaypointPathContainer().end())
+    {
+        for (WaypointPath::const_iterator it = itr->second.begin(); it != itr->second.end(); ++it)
+        {
+            if ((*it)->id == id)
+            {
+                itr->second.erase(it);
+                break;
+            }
+        }
+    }
+
+    Trinity::NormalizeMapCoord(x);
+    Trinity::NormalizeMapCoord(y);
+
+    WaypointPath& path = sWaypointMgr->GetWaypointPathContainer()[entry];
```

Comment:
The only way to defend TC branches from thread-unsafe code would be to have a helgrind bot...